src/inspector/main_thread_interface.cc

```diff
@@ -0,0 +1,283 @@
+  std::shared_ptr<CrossThreadInspectorSessionBlock> block_;
+};
+
+class ThreadSafeDelegate : public InspectorSessionDelegate {
+ public:
+  ThreadSafeDelegate(std::shared_ptr<MainThreadHandle> thread,
+                     std::unique_ptr<InspectorSessionDelegate> delegate)
+                     : thread_(thread), delegate_(std::move(delegate)) {}
+
+  ~ThreadSafeDelegate() override {
+    DeleteOnAnotherThread(thread_, std::move(delegate_));
+  }
+
+  void SendMessageToFrontend(const v8_inspector::StringView& message) override {
+    thread_->Post(std::unique_ptr<Request>(
+        new PostMessageRequest(delegate_, message)));
+  }
+
+ private:
+  std::shared_ptr<MainThreadHandle> thread_;
+  std::shared_ptr<InspectorSessionDelegate> delegate_;
+};
+}  // namespace
+
+
+MainThreadInterface::MainThreadInterface(Agent* agent, uv_loop_t* loop,
+                                         v8::Isolate* isolate,
+                                         v8::Platform* platform)
+                                         : agent_(agent), isolate_(isolate),
+                                           platform_(platform) {
+  main_thread_request_.reset(new AsyncAndInterface(uv_async_t(), this));
+  CHECK_EQ(0, uv_async_init(loop, &main_thread_request_->first,
+                            DispatchMessagesAsyncCallback));
+  // Inspector uv_async_t should not prevent main loop shutdown.
+  uv_unref(reinterpret_cast<uv_handle_t*>(&main_thread_request_->first));
+}
+
+MainThreadInterface::~MainThreadInterface() {
+  if (handle_)
+    handle_->Reset();
+}
+
+// static
+void MainThreadInterface::DispatchMessagesAsyncCallback(uv_async_t* async) {
+  AsyncAndInterface* asyncAndInterface =
+      node::ContainerOf(&AsyncAndInterface::first, async);
+  asyncAndInterface->second->DispatchMessages();
+}
+
+// static
+void MainThreadInterface::CloseAsync(AsyncAndInterface* pair) {
+  uv_close(reinterpret_cast<uv_handle_t*>(&pair->first), DisposePairCallback);
+}
+
+void MainThreadInterface::Post(std::unique_ptr<Request> request) {
+  Mutex::ScopedLock scoped_lock(state_lock_);
+  bool needs_notify = requests_.empty();
+  requests_.push_back(std::move(request));
+  if (needs_notify) {
+    CHECK_EQ(0, uv_async_send(&main_thread_request_->first));
+    if (isolate_ != nullptr && platform_ != nullptr) {
+      platform_->CallOnForegroundThread(isolate_,
+                                        new DispatchMessagesTask(this));
+      isolate_->RequestInterrupt(DispatchFromInterrupt, this);
+    }
+  }
+  incoming_message_cond_.Broadcast(scoped_lock);
+}
+
+bool MainThreadInterface::WaitForFrontendEvent() {
+  // We allow DispatchMessages reentry as we enter the pause. This is important
+  // to support debugging the code invoked by an inspector call, such
+  // as Runtime.evaluate
+  dispatching_messages_ = false;
+  if (dispatching_message_queue_.empty() && requests_.empty())
+    incoming_message_cond_.Wait(Mutex::ScopedLock(state_lock_));
```

Comment:
This doesn’t look quite right. `requests_` is a field that’s protected by `state_lock_`, right?

That means that the mutex needs to be locked outside of the conditional, before the `requests_.empty()` check (unless there are other synchronization primitives in place?)

Also, libuv doesn’t protect against spurious wakeups as far as I can tell, meaning that you want to turn the preceding `if (…)` into a `while (…)` loop.

Reply:
Done.